src/main/kotlin/io/github/inductiveautomation/kindling/log/LogTree.kt

```diff
@@ -0,0 +1,108 @@
+package io.github.inductiveautomation.kindling.log
+
+import com.formdev.flatlaf.extras.FlatSVGIcon
+import com.jidesoft.swing.CheckBoxTree
+import io.github.inductiveautomation.kindling.idb.metrics.Metric
+import io.github.inductiveautomation.kindling.idb.metrics.MetricNode
+import io.github.inductiveautomation.kindling.idb.metrics.MetricTree
+import io.github.inductiveautomation.kindling.idb.metrics.RootNode
+import io.github.inductiveautomation.kindling.utils.AbstractTreeNode
+import io.github.inductiveautomation.kindling.utils.TypedTreeNode
+import io.github.inductiveautomation.kindling.utils.asActionIcon
+import io.github.inductiveautomation.kindling.utils.treeCellRenderer
+import org.apache.commons.math3.stat.Frequency
+import javax.swing.tree.DefaultTreeModel
+import javax.swing.tree.TreeNode
+import javax.swing.tree.TreePath
+
+class LogEventNode(
+    override val userObject: List<String>,
+    frequency: Int = 0
+) : TypedTreeNode<List<String>>() {
+    //constructor(vararg parts: String) : this(parts.toList())
+
+    val name by lazy { userObject.joinToString(".") }
+
+    val frequency: Int by lazy {
+        if (this.isLeaf) {
+            frequency
+        } else {
+            //sum up all children of this node
+            this.children.sumOf {
+                (it as LogEventNode).frequency
+            }
+        }
+    }
+}
+
+class RootNode(logEvents: List<SystemLogEvent>) : AbstractTreeNode() {
+    private val logEventsByLogger : Map<String, Int> = logEvents.groupBy { it.logger }.mapValues { mapEntry ->
+        mapEntry.value.size
+    }
+
+    init {
+        val seen = mutableMapOf<List<String>, LogEventNode>()
+        for ((logger, freq) in logEventsByLogger.entries) {
+            var lastSeen: AbstractTreeNode = this
+            val currentLeadingPath = mutableListOf<String>("")
+            val loggerParts = logger.split('.')
+            for ((index, part) in loggerParts.withIndex()) {
+                currentLeadingPath.add(part)
+                val next = seen.getOrPut(currentLeadingPath.toList()) {
+                    val path = currentLeadingPath.drop(1)
+                    val newChild = if (index == loggerParts.size-1) {
+                        val name = path.joinToString(".")
+                        LogEventNode(path, logEventsByLogger[name]!!)
```

Comment:
This can be optimized by just using the `logger` and `freq` references in the main outer for-loop. Don't know how I didn't catch this before.